File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        HashMap<Integer,Integer> sum = new HashMap<>();
        sum.put(0,1);
        int count =0,totalSum =0;
        for(int num:nums)
        {
            totalSum+=num;
            if(sum.containsKey(totalSum-k))
            {
                count+=sum.get(totalSum-k);
            }
            sum.put(totalSum,sum.getOrDefault(totalSum,0)+1);
        }
    
    return count;
}
}

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
// TC =O(n)  SC = O(n)
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int,int> prefixsum;
        // we r starting from sum as 0 , so we need to enter it into hashtable.
        // if first element itself comes up as k , then we need to consider that.
        prefixsum.insert({0,1});
        int sum =0 ,count =0;
        for(int i =0;i<nums.size();i++)
        {
            sum+=nums[i];
            // if sum-k is already present , we need to take its value
             auto it = prefixsum.find(sum-k);
            if(it!=prefixsum.end())
            {
                count+=it->second;
            }
           
            // we have to find wether the current  sum already exists in the hashtable.
            // if exists , increment its count , if not insert it as new 
           it = prefixsum.find(sum);
            if(it!=prefixsum.end())
            {
                it->second++;
            }
            else
            prefixsum.insert({sum,1});
             


        }
        return count;
    }
};